fs_impl.go

```diff
@@ -0,0 +1,353 @@
+}
+
+func (fs *Fs) touch(parent_uuid uuid.UUID, name string) (uuid.UUID, error) {
+	return fs.newRecord(parent_uuid, name)
+}
+
+func (fs *Fs) delete(parent uuid.UUID, child uuid.UUID) error {
+	err := fs.deleteChildInMemoryAndWriteParentToDisk(parent, child)
+	if err != nil {
+		return err
+	}
+
+	err = fs.deleteRecordFilesFromDisk(child.String())
+	if err != nil {
+		return err
+	}
+
+	return fs.decreseChildrenRefCountRecursive(parent)
+}
+
+func (fs *Fs) mount(parent uuid.UUID, newChild uuid.UUID) error {
+	rec := fs.records[parent]
+
+	for _, child := range rec.Children {
+		if child == newChild {
+			return errors.New("child with this uuid already exists")
+		}
+	}
+
+	rec.Children = append(rec.Children, newChild)
+
+	fs.records[newChild].refs++
+
+	return fs.writeToDisk(parent, rec)
+}
+
+func (fs *Fs) readSection(uuid uuid.UUID, section string) ([]byte, error) {
+	err := checkSectionNameIsSane(section)
+	if err != nil {
+		return nil, err
+	}
+	return os.ReadFile(fs.getSectionFileName(uuid, section))
+}
+
+func (fs *Fs) writeSection(uuid uuid.UUID, section string, data []byte) error {
+	err := checkSectionNameIsSane(section)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(fs.getSectionFileName(uuid, section), data, 0644)
+}
+
+func (fs *Fs) deleteSection(uuid uuid.UUID, section string) error {
+	return os.Remove(fs.getSectionFileName(uuid, section))
+}
+
+//---
+
+func (fs *Fs) loadRecords() error {
+	entries, err := os.ReadDir(fs.basePath)
+	if err != nil {
+		return err
+	}
+
+	var sectionFiles []string
+	var recordFiles []string
+
+	for _, e := range entries {
+		if e.Type().IsDir() {
+			return errors.New("garbage directory in fs root")
+		}
+
+		name := e.Name()
+
+		match, _ := regexp.MatchString(onlyFileInFsRootPattern, name)
+		if !match {
+			return errors.New("garbage file in fs root")
+		}
+
+		if len(name) == 36 {
+			recordFiles = append(recordFiles, name)
+		} else {
+			sectionFiles = append(sectionFiles, name)
+		}
+	}
+
+	for _, recordName := range recordFiles {
+		u, err := uuid.Parse(recordName)
+		if err != nil {
+			return err
+		}
+
+		f, err := os.Open(fs.path(recordName))
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		rec := new(Record)
+		dec := json.NewDecoder(f)
+		err = dec.Decode(rec)
+		if err != nil {
+			return err
+		}
+
+		fs.records[u] = rec
+	}
+
+	// TODO(prokop) load section file names
+	return nil
+}
+
+func checkLoadedRecordsAreSane(map[uuid.UUID]*Record) error {
+	// TODO(prokop)
+	return nil
+}
+
+func newFs(root uuid.UUID, basePath string) (fs Fs, err error) {
+	fs.basePath = basePath
+	fs.root = root
+	fs.records = make(map[uuid.UUID]*Record)
+
+	err = fs.loadRecords()
+	if err != nil {
+		return
+	}
+
+	if _, c := fs.records[root]; !c {
+		err = errors.New("the root UUID not found in fs")
+		return
+	}
+
+	return fs, checkLoadedRecordsAreSane(fs.records)
+}
+
+func handleLs(userStore userStorer, fileStore fileStorer) http.Handler {
```

Comment:
I don't think this should be in this file.

Reply:
i agree. where do we put it?

Reply:
fs_endpoints.go